File: client/android/core/src/main/java/io/github/pstanar/pstotp/core/crypto/IconLibraryCrypto.kt
package io.github.pstanar.pstotp.core.crypto

import io.github.pstanar.pstotp.core.model.IconLibraryBlob

/**
 * Encrypt / decrypt the user's icon-library blob with the vault key.
 * Mirrors the web client's icon-library-crypto.ts byte-for-byte: same
 * AEAD (AES-256-GCM), same AAD ("pstotp-icon-library-v1"), same
 * nonce(12) || ciphertext+tag(16) layout. A blob written on one
 * platform must decrypt on the other.
 */
object IconLibraryCrypto {

    /** Fixed AD per blob-format version. Changing the string invalidates
     *  all existing blobs — only do that with an explicit migration. */
    private val AD_V1 = "pstotp-icon-library-v1".toByteArray(Charsets.UTF_8)

    /**
     * Encrypt a blob with the given vault key. Returns the raw
     * nonce||ciphertext||tag bytes; the caller base64-encodes before
     * sending to the server or storing locally.
     */
    fun encrypt(vaultKey: ByteArray, blob: IconLibraryBlob): ByteArray {
        val plaintext = blob.toJsonString().toByteArray(Charsets.UTF_8)
        return AesGcm.encrypt(vaultKey, plaintext, AD_V1)
    }

    /**
     * Decrypt a blob fetched from the server or loaded from local
     * settings. Throws on MAC failure / wrong key.
     */
    fun decrypt(vaultKey: ByteArray, payload: ByteArray): IconLibraryBlob {
        val plaintext = AesGcm.decrypt(vaultKey, payload, AD_V1)
        return IconLibraryBlob.fromJsonString(String(plaintext, Charsets.UTF_8))
    }
}

File: client/web/src/features/vault/utils/icon-library-crypto.ts
import { aesGcmEncrypt, aesGcmDecrypt, toBase64, fromBase64 } from "@/lib/crypto";
import type { IconLibraryBlob } from "@/types/icon-library";

const NONCE_LENGTH = 12;

/**
 * Associated data for the library blob's AEAD. A fixed string per
 * blob-format version: changes the AD if we ever need to rotate the
 * blob schema incompatibly, so an old blob can't be rewrapped as a new
 * one without explicit migration.
 */
const ICON_LIBRARY_AD_V1 = new TextEncoder().encode("pstotp-icon-library-v1");

/**
 * Encrypt the library JSON with the user's vault key. Produces the
 * same nonce(12) || ciphertext+tag layout the vault-entry blob uses,
 * base64-encoded for the wire.
 */
export async function encryptIconLibrary(
  vaultKey: Uint8Array,
  blob: IconLibraryBlob,
): Promise<string> {
  const json = new TextEncoder().encode(JSON.stringify(blob));
  const { ciphertext, nonce } = await aesGcmEncrypt(vaultKey, json, ICON_LIBRARY_AD_V1);
  const packed = new Uint8Array(nonce.length + ciphertext.length);
  packed.set(nonce, 0);
  packed.set(ciphertext, nonce.length);
  return toBase64(packed);
}

/**
 * Decrypt a library blob fetched from the server. Throws on MAC failure
 * / wrong key — caller should handle by treating the library as empty
 * (with a warning) rather than silently dropping it.
 */
export async function decryptIconLibrary(
  vaultKey: Uint8Array,
  base64Payload: string,
): Promise<IconLibraryBlob> {
  const packed = fromBase64(base64Payload);
  const nonce = packed.slice(0, NONCE_LENGTH);
  const ciphertextWithTag = packed.slice(NONCE_LENGTH);
  const json = await aesGcmDecrypt(vaultKey, ciphertextWithTag, nonce, ICON_LIBRARY_AD_V1);
  return JSON.parse(new TextDecoder().decode(json)) as IconLibraryBlob;
}
